modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -156,19 +171,31 @@ public void triggered(SchedulerEngine.Event event) {
     // default visibility for testing purposes
     void run(ClusterState state) {
         for (DataStream dataStream : state.metadata().dataStreams().values()) {
+            clearErrorStoreForUnmanagedIndices(dataStream);
             if (dataStream.getLifecycle() == null) {
                 continue;
             }
 
+            String writeIndex = dataStream.getWriteIndex().getName();
             try {
                 maybeExecuteRollover(state, dataStream);
             } catch (Exception e) {
                 logger.error(() -> String.format(Locale.ROOT, "DLM failed to rollver data stream [%s]", dataStream.getName()), e);
+                DataStream latestDataStream = clusterService.state().metadata().dataStreams().get(dataStream.getName());
+                if (latestDataStream != null) {
+                    if (latestDataStream.getWriteIndex().getName().equals(writeIndex)) {
+                        // data stream has not been rolled over in the meantime so record the error against the write index we
+                        // attempted the rollover
+                        errorStore.recordError(writeIndex, e);
+                    }
+                }
             }
 
             try {
                 maybeExecuteRetention(state, dataStream);
             } catch (Exception e) {
+                // individual index errors would be reported via the API action listener for every delete call
+                // we could potentially record errors at a data stream level and expose it via the _data_stream API?
```

Comment:
It seems like we would want to store these failures as well right? For instance, if the deletion couldn't occur because the filesystem was read-only